demos/device_defender_for_aws/defender_demo.c

```diff
@@ -427,6 +433,81 @@ static BaseType_t collectDeviceMetrics( void )
         }
     }
 
+    if( metricsCollectorStatus == MetricsCollectorSuccess )
+    {
+        /* Get task count */
+        arraySize = uxTaskGetNumberOfTasks();
+
+        /* Allocate pTaskStatusArray */
+        pTaskStatusArray = pvPortMalloc( arraySize * sizeof( TaskStatus_t ) );
+
+        if( pTaskStatusArray == NULL )
+        {
+            LogError( ( "Allocating pTaskStatusArray failed." ) );
+            metricsCollectorStatus = MetricsCollectorCollectionFailed;
+        }
+    }
+
+    if( metricsCollectorStatus == MetricsCollectorSuccess )
+    {
+        /* Allocate pTaskIdArray */
+        pTaskIdArray = pvPortMalloc( arraySize * sizeof( uint32_t ) );
+
+        if( pTaskIdArray == NULL )
+        {
+            LogError( ( "Allocating pTaskIdArray failed." ) );
+            metricsCollectorStatus = MetricsCollectorCollectionFailed;
+        }
+    }
+
+    /* Collect custom metrics from the system to send to AWS IoT Device Defender.
+     * This demo sends this task's stack high water mark as a "number" type
+     * custom metric and the current task ids as a "list of numbers" type custom
+     * metric. */
+    if( metricsCollectorStatus == MetricsCollectorSuccess )
+    {
+        /* Get the current task's status information. The usStackHighWaterMark
+         * field of the task status will be included in the report as a "number"
+         * custom metric. */
+        vTaskGetInfo(
+            /* NULL has the function query this task. */
+            NULL,
+            &taskStatus,
+            /* Include the stack high water mark value. */
+            pdTRUE,
+            /* Don't include the task state in the TaskStatus_t structure. */
+            0 );
+
+        /* Get the task status information for all running tasks. The task IDs
+         * of each task is then extracted to include in the report as a "list of
+         * numbers" custom metric */
+        tasksWritten = uxTaskGetSystemState( pTaskStatusArray, arraySize, NULL );
+
+        if( tasksWritten == 0 )
+        {
+            /* If 0 is returned, the buffer was too small */
+            metricsCollectorStatus = MetricsCollectorCollectionFailed;
+            LogError( ( "Failed to collect task IDs. uxTaskGetSystemState() failed due to insufficient buffer space." ) );
+        }
+        else
+        {
+            size_t i;
+
+            /* Populate the task IDs array from the collected system state array. */
+            for( i = 0; i < tasksWritten; i++ )
+            {
+                pTaskIdArray[ i ] = pTaskStatusArray[ i ].xTaskNumber;
+            }
+        }
+    }
+
+    /* Free old pTaskIdsArray */
+    if( deviceMetrics.pTaskIdsArray != NULL )
+    {
+        vPortFree( deviceMetrics.pTaskIdsArray );
+        deviceMetrics.pTaskIdsArray = NULL;
+    }
```

Comment:
Why is freeing the `deviceMetrics.pTaskIdsArray` required? Isn't the `pTaskIdArray` array being freed later in this function anyway?

Reply:
pTaskIdArray is freed later if it is not added to the struct. If it is added to the struct, it survives past the function, so we clean it up before we replace it. Else we would leak memory when we set the struct member to the new pointer.

Reply:
Why not always free the array irrespective of whether it is added to the metrics struct or not? Wouldn't the memory leak if the demo loop is only executed once with this approach?

Reply:
If the array is not added to the metrics struct, it is freed at end of the function. That free is because when the array *is* added to the struct, it needs to be cleaned. Moved the success case free to the main loop though.